gcc/rust/lex/rust-lex.h

```diff
@@ -208,37 +207,136 @@ class Lexer
   Linemap *get_line_map () { return line_map; }
   std::string get_filename () { return std::string (input.get_filename ()); }
 
-private:
-  void start_line (int current_line, int current_column);
-
-  // File for use as input.
-  RAIIFile input;
-  // TODO is this actually required? could just have file storage in InputSource
+  // Input source wrapper thing.
+  class InputSource
+  {
+  private:
+    // position of current character
+    unsigned int pos;
+    std::vector<Codepoint> chars;
+    bool is_valid_utf8;
```

Comment:
`is_valid_utf8`represents whether input string is valid as utf-8 and it is properly decoded by this instance. (Decoded characters are buffered in `chars` field)

Reply:
So we have to stop the compilation pipeline if this field is set to false.

Reply:
Yeah that should be a separate PR